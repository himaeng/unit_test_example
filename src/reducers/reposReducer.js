import {
  LOAD_REPOS_REQUEST,
  LOAD_REPOS_SUCCESS,
  LOAD_REPOS_FAILURE,
  SELECT_REPO
} from '../actions/types';

export const initialState = {
  error: null,
  isLoading: false,
  repos: null,
  selected: null
}

export default function repos(state = initialState, action) {
  switch (action.type) {
    case LOAD_REPOS_FAILURE:
      return { ...state, error: action.error, isLoading: false };
    case LOAD_REPOS_REQUEST:
      return { ...state, isLoading: true };
    case LOAD_REPOS_SUCCESS:
      return { ...state, isLoading: false, repos: action.data }
    case SELECT_REPO:
      const selected = state.selected !== action.id ? action.id : null;
      return { ...state, selected };
    default:
      return state;
  }
}
