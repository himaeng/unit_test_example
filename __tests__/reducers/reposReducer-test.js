import reposReducer,
{
  initialState
} from '../../src/reducers/reposReducer';
import {
  loadReposRequest,
  loadReposSuccess,
  loadReposFailure,
  selectRepo
} from '../../src/actions';
import { repos as data } from '../../config/jest/mockData';

describe('reposReducer', () => {
  it('return initialState หากไม่เจอ action ที่ตรงกัน', () => {
    expect(reposReducer(initialState, {type: 'มั่วๆ'})).toMatchSnapshot();
  });

  it('จัดการกับเคส failure action', () => {
    const error = new Error('ดูผมดิ ผมเออเร่อ');
    expect(reposReducer(initialState, loadReposFailure(error))).toMatchSnapshot();
  });

  it('จัดการกับเคส request action', () => {
    expect(reposReducer(initialState, loadReposRequest())).toMatchSnapshot();
  });

  it('จัดการกับเคส success action', () => {
    expect(reposReducer(initialState, loadReposSuccess(data))).toEqual({
      ...initialState,
      repos: data,
      isLoading: false
    });
  });

  it('จัดการกับเคส success action แบบ snapshot', () => {
    expect(reposReducer(initialState, loadReposSuccess(data))).toMatchSnapshot();
  });

  it('จัดการกับเคส SELECT_REPO', () => {
    expect(reposReducer(initialState, selectRepo(1000))).toMatchSnapshot();
  });

  it('จัดการกับเคส ยิง undefined state กับ unknow action', () => {
    expect(reposReducer(undefined, { type: 'unknown'})).toMatchSnapshot();
  });
});
