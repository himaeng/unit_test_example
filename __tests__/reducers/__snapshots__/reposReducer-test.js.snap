// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reposReducer return initialState หากไม่เจอ action ที่ตรงกัน 1`] = `
Object {
  "error": null,
  "isLoading": false,
  "repos": null,
  "selected": null,
}
`;

exports[`reposReducer จัดการกับเคส SELECT_REPO 1`] = `
Object {
  "error": null,
  "isLoading": false,
  "repos": null,
  "selected": 1000,
}
`;

exports[`reposReducer จัดการกับเคส failure action 1`] = `
Object {
  "error": [Error: ดูผมดิ ผมเออเร่อ],
  "isLoading": false,
  "repos": null,
  "selected": null,
}
`;

exports[`reposReducer จัดการกับเคส request action 1`] = `
Object {
  "error": null,
  "isLoading": true,
  "repos": null,
  "selected": null,
}
`;

exports[`reposReducer จัดการกับเคส success action แบบ snapshot 1`] = `
Object {
  "error": null,
  "isLoading": false,
  "repos": Array [
    Object {
      "description": "Useful description goes here.",
      "id": 1,
      "name": "Test 1",
      "stargazers_count": 50000,
    },
    Object {
      "description": "Useful description goes here.",
      "id": 2,
      "name": "Test 2",
      "stargazers_count": 10000,
    },
    Object {
      "description": "Useful description goes here.",
      "id": 3,
      "name": "Test 3",
      "stargazers_count": 499,
    },
  ],
  "selected": null,
}
`;

exports[`reposReducer จัดการกับเคส ยิง undefined state กับ unknow action 1`] = `
Object {
  "error": null,
  "isLoading": false,
  "repos": null,
  "selected": null,
}
`;
